Step 1b: Calculating cell-trait association scores
================

This is an example of the analysis step of calculating cell-trait
association scores. We use the single-cell RNA-seq dataset of pancreatic
islet development from [Byrnes et
al](https://doi.org/10.1038/s41467-018-06176-3).

Download the Monocle object ‘E14\_fev\_lineage\_monocle\_ob.Rdata’ from
figshare.com:
[link](https://figshare.com/articles/dataset/Monocle_Objects_-_V2_Dataset/6783554?backTo=/collections/Lineage_dynamics_of_murine_pancreatic_development_at_single-cell_resolution/4158458).

Load the following required packages:

``` r
library(monocle)
library(Seurat)
library(data.table)
```

Download functions from this github directory
[here](https://github.com/ElkonLab/scGWAS/blob/master/R/functions_scGWAS.R),
and source:

``` r
source('functions_GWAS_traj.R')
```

### Preprocessing

We will use extract the count matrix from the monocle object. Fir, Load
the object:

``` r
load('E14_fev_lineage_monocle_ob.Rdata')
```

The count matrix is obtained by:

``` r
exp <- HSMM_seur_var@assayData$exprs
```

To save working memory, it might be a good idea to remove the monocle
object from the environment.

``` r
rm('HSMM_seur_var')
```

We keep genes expressed in at least 10 cells:

``` r
genes.keep <- rowSums(as.matrix(exp) > 0) >= 10
exp <- exp[genes.keep,]
```

Normalize the data using Seurat’s log-normalization:

``` r
exp <- Seurat::NormalizeData(exp)
```

Last, we convert the mouse genes to human orthologues. We provid a
conversion table for ‘one to one’ orthologs
[here](https://github.com/ElkonLab/scGWAS/blob/master/data/rds/ortholog_one2one.RDS).
Other tables can be obtained from
[BioMart](https://www.ensembl.org/biomart/martview/68453bf127464cacc5a5d064c92158e0).

``` r
ortholog_one2one <- readRDS("ortholog_one2one.RDS")
head(ortholog_one2one, 3)
# convert the expression matrix into a data.frame
exp <- cbind.data.frame(data.frame(GENE = rownames(exp)), as.data.frame(exp))
# merge
exp <- merge(ortholog_one2one, exp, by = "GENE")
exp <- exp[,-2]
exp <- exp[!duplicated(exp$GENE),]
```

#### Calculating cell-trait association scores

This part involves running MAGMA’s gene property analysis for each cell
separately. This might take long, so consider using multiple cores.
First, we generate ‘covariate files’ for MAGMA for each cell. Each
output file is a table with the columns: gene, normalized expression in
the cell, and average normalized expression in the dataset. The function
takes as input the data.frame of the normalized expression matrix. The
user specifies the name of the output directory (out.dir) and the number
of cores (cor).

``` r
generate_covs_cells(exp = exp, out.dir = "cov_panc",cors = 20)
```

The next step, it running the gene property analysis, for each cell.
This analysis firs the following the regression model to each cell:

![](https://github.com/eldadshulman/scGWAS/blob/master/data/pic/eq.PNG)

where Z is the vector of the gene’s Z-score converted from the p-values
obtained from MAGMA’s gene analysis for the trait. B is a matrix of
technical confounders, including gene length and SNPs’ LDs, calculated
by MAGMA. C is the vector of normalized expression of the genes in the
cell, and A is a vector of the average normalized expression for the
gene in the dataset. The t-statistic of *β*<sub>*c*</sub> is taken as
the score for the association between the cell and trait.

This function will run the regression for each cell. The input is:

-   magma.path - The path to magma directory
    (“/path/to/MAGMA\_directory/”)

-   covariate.files.path - The path to the covariate files directory
    created by the previews step.

-   raw.file.path - The path to the output (raw) of MAGMA’s gene
    analysis for the trait. We provide the output
    [here](https://github.com/ElkonLab/scGWAS/blob/master/data/magma_outputs/T2D_European.BMIadjusted.genes.raw).

-   out.dir - The name of the output directory.

-   cor - the number of cores to use.

We will get MAGMA gene property output files for each cell. Note that
this step takes even longer.

``` r
gene_prop_cells(cor = 30, raw.file.path = "T2D_European.BMIadjusted.genes.raw", 
                covariate.files.path = "cov_panc", out.dir = "T2D")
```

Last, extract the cell scores from MAGMA’s output, namely the
t-statistics.

``` r
cs.pan.t2d <- get_scores(gsa.files.path = "T2D", cors = 20)
```

``` r
head(cs.pan.t2d)
```

    ##              cells         cs
    ## 1 AAACCTGAGAGAGCTC  0.4663375
    ## 2 AAACCTGAGAGTGACC -2.0146530
    ## 3 AAACCTGCATGTAGTC  1.3549249
    ## 4 AAACCTGCATTCACTT -0.9757387
    ## 5 AAACCTGGTAGAAGGA -0.7142417
    ## 6 AAACCTGGTTACTGAC -0.4747647

The first column is cell barcode and the second is the cell score (T2D
risk association score).

``` r
saveRDS(object = cs.pan.t2d, file = "cs.pan.t2d.RDS")
```
