Step 2a: Elucidate molecular pathways that underlie the link between the
trajectory and trait
================

In this step we will identify gene sets that are enriched in both the
trajectory of pancreas development and trait (T2D).

### Load packeges

Load the following required packages:

``` r
library(msigdbr)
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)
library(dplyr)
```

Download functions from this github directory
[here](https://github.com/ElkonLab/scGWAS/blob/master/R/functions_scGWAS.R),
and source:

``` r
source('functions_GWAS_traj.R')
```

### Load data

From [**step
1b**](https://github.com/ElkonLab/scGWAS/blob/master/vignettes/1d.md),
we got a data.frame with the expression matrix normalized, filtered, and
with genes coverted to human orthologs. You can download the output
[here](https://github.com/ElkonLab/scGWAS/blob/master/data/rds/cs.pan.t2d.RDS).
We load it:

``` r
exp <- readRDS("cs.pan.t2d.RDS")
head(exp, 3)
```

    ##              cells         cs
    ## 1 AAACCTGAGAGAGCTC  0.4663375
    ## 2 AAACCTGAGAGTGACC -2.0146530
    ## 3 AAACCTGCATGTAGTC  1.3549249

From **step 1d**, we go the cell metadata. We load it.

``` r
meta <- readRDS("meta.pancreas.RDS")
head(exp, 3)
```

    ##              cells         cs
    ## 1 AAACCTGAGAGAGCTC  0.4663375
    ## 2 AAACCTGAGAGTGACC -2.0146530
    ## 3 AAACCTGCATGTAGTC  1.3549249

We load the C5 gene set (GO terms) from msigdbr R packege.

``` r
m_t2g <- msigdbr(species = "Homo sapiens", category = "C5") %>% 
  dplyr::select(gs_name, entrez_gene)
head(m_t2g,3)
```

    ## # A tibble: 3 x 2
    ##   gs_name                                         entrez_gene
    ##   <chr>                                                 <int>
    ## 1 GO_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY      390637
    ## 2 GO_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY        4507
    ## 3 GO_1_4_ALPHA_OLIGOGLUCAN_PHOSPHORYLASE_ACTIVITY        5834

### Ordering the genes according to pseudotime effect

First, we match the orde of the cells in the rows of the metadata, with
that of the expression matrix.

``` r
meta <- meta[match(colnames(exp), meta$cells),]
```

To use GSEA, we first created a ranked list for the genes in the
dataset, according to the incremental change in their expression along
the trajectory. We estimated these incremental changes using GLM to
model the effect of pseudotime on genesâ€™ expression. We use
log-normalized gene expression values and a Gaussian error distribution
GLM. The number of genes detected in each cell was added as a covariate
(nGene).

``` r
p.effect <- df.genes(exp = exp, meta = meta, model = '~Pseudotime:beta + nGene', brach = 'beta', cors = 10)
```

``` r
head(p.effect, 3)
```

    ##       gene   Estimate Std. Error       t Pr(>|t|)
    ## 1  C14orf1  1.404e-04  2.531e-03  0.0555 9.56e-01
    ## 2 KIAA0141  2.569e-03  1.949e-03  1.3180   0.1876
    ## 3 KIAA1841 -1.836e-02  2.173e-03 -8.4490 4.72e-17

### Running GSEA

To use GSEA from clusterProfiler, we covert gene symbols to ENTREZ IDs.

``` r
  gene.df <- bitr(p.effect$SYMBOL, fromType = "SYMBOL" ,
                  toType = c("ENTREZID"),
                  OrgDb = org.Hs.eg.db)
  p.effect <- merge(p.effect, gene.df, by = "SYMBOL")
```

We use the t-statistics of the pseudotime:branch coefficients for
ranking the genes.

``` r
  p.effect <- dplyr::arrange(p.effect, dplyr::desc(t))
  geneList <- as.numeric(p.effect$t)
  names(geneList) <- p.effect$gene
```

``` r
 head(geneList, 3)
```

    ##    5126    3375    7276 
    ## 39.7417 39.3662 36.5741

``` r
tail(geneList,3)
```

    ##     2877     6142    50674 
    ## -35.0638 -38.3016 -53.2655

We run GSEA using the ontology gene set (GO, C5, 50 &lt; set size &lt;
500) from MSigDB.

``` r
  edo <-  clusterProfiler::GSEA(geneList = geneList, TERM2GENE =m_t2g, maxGSSize = 500, 
              minGSSize = 50, pvalueCutoff = 1, nPerm = 100000)
  edo <- setReadable(x = edo, OrgDb = org.Hs.eg.db, keyType = "ENTREZID")
```

``` r
 head(edo,2)
```

    ##                                                                                                        ID
    ## GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION
    ## GO_CELL_DIVISION                                                                         GO_CELL_DIVISION
    ##                                                                                               Description
    ## GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION
    ## GO_CELL_DIVISION                                                                         GO_CELL_DIVISION
    ##                                                      setSize enrichmentScore
    ## GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION     495      -0.4172673
    ## GO_CELL_DIVISION                                         492      -0.4068760
    ##                                                            NES       pvalue
    ## GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION -1.659186 1.286389e-05
    ## GO_CELL_DIVISION                                     -1.617105 1.288012e-05
    ##                                                          p.adjust      qvalues
    ## GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION 0.0008616726 0.0006815505
    ## GO_CELL_DIVISION                                     0.0008616726 0.0006815505
    ##                                                      rank
    ## GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION 2245
    ## GO_CELL_DIVISION                                     2896
    ##                                                                        leading_edge
    ## GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION tags=26%, list=18%, signal=22%
    ## GO_CELL_DIVISION                                     tags=42%, list=23%, signal=33%
    ##                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    core_enrichment
    ## GO_POSTTRANSCRIPTIONAL_REGULATION_OF_GENE_EXPRESSION                                                                                                                                                                                                                                                                                                                                                                                                                         MEX3D/NCOR2/UNK/RBMS3/C1QBP/EIF4B/PSMD8/HIST1H3F/RBM24/ZFP36L1/ILF3/ROCK1/UPF3A/HNRNPLL/PSMD12/PSMA7/DKC1/PSMA4/NEURL1/HABP4/PSMD5/POLR2C/NUP85/NEMF/DDX1/ATF4/AGO1/EIF4H/SECISBP2/ELAVL1/ENC1/ZC3H14/PIWIL4/CNOT6/AAAS/PIWIL1/MYCN/TYMS/RBM10/CDC37/EIF3K/NANOS1/LSM14A/PSMC3/MAP2K1/IGF2BP2/SMAD1/PINK1/TAF15/GRB7/FMR1/CNOT2/SYNCRIP/MIF4GD/MTPN/PSMC4/TNRC6B/EIF2S1/ERBB2/GTPBP2/TGFB1/H3F3B/ATG14/HIST2H3A/EIF2B2/PSMB1/APP/CNOT7/ZFP36/PAIP2/PSMD7/APEX1/CNBP/RPL30/PAIP1/DNAJC1/KHDRBS1/EIF4EBP2/YTHDC1/PSMD11/PSMA2/FXR1/MAPK3/DDX39B/E2F1/RNASEL/HIST2H4B/EIF4E3/PSMD10/YTHDF2/SET/TSN/EIF1/EIF2A/NUDT21/HNRNPD/HSPA8/SERBP1/EIF5B/TP53/PSMC2/CNOT6L/MAGOH/PSMD4/TOB1/RPS27A/HNRNPA2B1/PSMC6/NDC1/HNRNPM/ELAVL4/VIM/PPP3CA/EIF3H/DDX5/EEF2/RPL5/UBA52/RPL26/EIF3E/GSPT1/CDK4/RPS9/RPS14/SOX4/RPS4X/BTG2/RPS3
    ## GO_CELL_DIVISION                                     STAMBP/WWTR1/TRIOBP/MAU2/HAUS1/CCNE2/VPS4B/PTCH1/RBL1/E2F8/NEDD1/MASTL/CNTROB/SKA3/CDK7/MAP9/CCNO/SPAG5/BORA/PLK4/CEP63/SIRT2/KATNA1/ERCC6L/NSMCE2/MLLT3/ZW10/GPSM2/FBXO5/RAB35/CCNF/DSN1/RACGAP1/KIF14/ANAPC7/WEE1/USP39/ANAPC4/NCAPH/WASL/CIB1/PLK2/PSRC1/CDC16/NUP62/CKAP2/DIXDC1/NCAPD3/BUB3/CDC6/CHFR/ARPP19/RCC2/ROCK1/SPDL1/FGF9/ECT2/ARHGEF2/MAEA/FGFR2/HAUS8/BRCC3/E2F7/BUB1B/FAM83D/SPICE1/SPAST/ASPM/SMC3/PIK3C3/SNX9/NCAPG2/ZNF830/NCAPD2/FIGN/PKN2/CDK1/SMC4/ZNF207/IL1B/THOC2/SAC3D1/CCND1/BUB1/INCENP/LIG1/CENPW/MAD2L1/PARD3/CDC25C/PLK1/RBBP8/TEAD3/MCMBP/EVI5/CEP55/OIP5/CDCA8/CDCA5/BCL2L1/KIF2A/ZWILCH/CENPJ/HAUS7/SKA1/PTN/NEK2/AKNA/NCAPG/AURKB/CDCA2/LZTS2/ETV5/PPP2R2D/VANGL2/MIS18BP1/HAUS3/TGFB1/SAPCD2/CDCA3/NUF2/TSG101/AURKA/MAD2L2/NEK6/CHMP4B/CCNA2/BIRC5/TOP2A/TERF1/PPP1CB/KIF20B/FGFR1/UBE2C/KIF2C/UBE2S/TIMELESS/PPP1CC/TXNL4A/TACC3/CCND3/CDC42/SPC24/KIF23/TIPIN/SMC1A/NDE1/LBH/CCNB2/HMGA2/KIF20A/REEP3/CCNB1/DLL1/KNSTRN/CDC26/SEPT3/TTC28/RALA/NDC80/KIF11/CDC20/PRC1/TPX2/NUSAP1/SMC2/RHOB/CECR2/CENPE/CALM1/CENPF/CDK2/SH3GLB1/HELLS/RTKN/CAT/CKS2/REEP4/SYCE2/IGF2/STOX1/SDE2/RHOC/PGF/TOP1/PRPF40A/CENPA/SEPT7/SFN/CFL1/SON/PLK5/KIT/NUMBL/CLTA/CDC14B/TXNIP/CCNG2/UBE2I/PLK3/CDK4/STMN1/MDK/RPS3

We can plot

``` r
enrichplot::gseaplot2(edo, geneSetID = "GO_REGULATION_OF_INSULIN_SECRETION", pvalue_table = T)
```

![](https://github.com/ElkonLab/scGWAS/blob/master/data/pic/GSEA.png)

Notice that in our manuscript, the plots are inverted. There we choose
rank genes in an increasing rather than decreasing order. We did it so
that the pseudotime direction is kept from left to right. Therefore,
while in the manuscript we sought for enrichment score (ES) &lt; 0, here
we get the same results from ES &gt; 0.
