Step 3: Prioritize genes that carry the link between the pathway, trait
and trajectory
================

In this step we will identify gene sets that are enriched in both the
trajectory of pancreas development.

### Load packeges

Load the following required packages:

``` r
library(msigdbr)
library(clusterProfiler)
library(org.Hs.eg.db)
library(enrichplot)
library(dplyr)
```

Download functions from this github directory
[here](https://github.com/ElkonLab/scGWAS/blob/master/R/functions_scGWAS.R),
and source:

``` r
source('functions_GWAS_traj.R')
```

### Load data

From **step 1a**, we got an output file (“gene.out”) from MAGMA’s gene
analysis, giving gene-trait (T2D) association p-values. It can be
downloaded from
[here](https://github.com/ElkonLab/scGWAS/blob/master/data/magma_outputs/T2D_European.BMIadjusted.genes.out).
We load it:

``` r
gene.p <- read.table("T2D_European.BMIadjusted.genes.out", header = T)
head(gene.p, 3)
```

    ##     GENE CHR  START   STOP NSNPS NPARAM       N   ZSTAT       P
    ## 1 SAMD11   1 859993 879961    13      8 1740000 0.20813 0.41756
    ## 2  NOC2L   1 879583 894679    19     13 1740000 0.97740 0.16419
    ## 3 KLHL17   1 895967 901099    19     13 1740000 0.40072 0.34431

From [**step
2b**](https://github.com/ElkonLab/scGWAS/blob/master/vignettes/2b.md),
we got the output from MAGMA’s gene-set analysis. This file has T2D risk
association p-values for the leading-edge gene sets enriched in the
pancreas trajectory. It can be downloaded from
[here](https://github.com/ElkonLab/scGWAS/blob/master/data/magma_outputs/t2d.panc.gsa.out).
We load it:

``` r
gsa <- read.table("t2d.panc.gsa.out", header = T)

head(gsa, 3)
```

                             VARIABLE TYPE NGENES      BETA   BETA_STD      SE
    1 GO_ACTIVE_TRANSMEMBRANE_TRAN...  SET     33  0.144420  0.0064990 0.18230
    2 GO_ADENYLATE_CYCLASE_ACTIVAT...  SET      8  0.099064  0.0021967 0.35044
    3        GO_ATP_METABOLIC_PROCESS  SET     48 -0.056490 -0.0030646 0.13070
            P
    1 0.21413
    2 0.38871
    3 0.66720
                                                                         FULL_NAME
    1                                 GO_ACTIVE_TRANSMEMBRANE_TRANSPORTER_ACTIVITY
    2 GO_ADENYLATE_CYCLASE_ACTIVATING_G_PROTEIN_COUPLED_RECEPTOR_SIGNALING_PATHWAY
    3                                                     GO_ATP_METABOLIC_PROCESS

From the same step, we also got a file with the trajectory significant
gene-sets and their leading-edge genes.

``` r
set.annot <- read.table("panc.leading-edge.txt", header = F)
colnames(set.annot) <- c("Gene-set", "GENE")
head(set.annot, 3)
```

                 Gene-set GENE
    1 GO_HORMONE_ACTIVITY IAPP
    2 GO_HORMONE_ACTIVITY  TTR
    3 GO_HORMONE_ACTIVITY  PYY

### Processing the data

First, we keep gene-sets significantly associated with T2D:

``` r
gsa <- gsa[which(gsa$P < 0.05),]

sig.gene.sets <- gsa$FULL_NAME
```

Now, we keep the genes from the leading-edge of the significant
(trajectory and trait) gene sets:

``` r
risk.genes <- set.annot[which(set.annot$`Gene-set` %in% sig.gene.sets),]
```

Many genes are members of more than one set and therefore appear
multiple times in this table. We use dplyr to collapse the rows of such
genes as follows:

``` r
risk.genes <- dplyr::group_by(risk.genes, GENE) %>% dplyr::summarise(`Gene-sets` = paste(`Gene-set`, collapse = ", "))
```

    `summarise()` ungrouping output (override with `.groups` argument)

``` r
head(risk.genes,3)
```

    # A tibble: 3 x 2
      GENE  `Gene-sets`                                                             
      <chr> <chr>                                                                   
    1 AACS  GO_HORMONE_TRANSPORT, GO_REGULATION_OF_PEPTIDE_SECRETION, GO_REGULATION~
    2 ABAT  GO_HORMONE_TRANSPORT, GO_REGULATION_OF_PEPTIDE_SECRETION, GO_REGULATION~
    3 ABCA2 GO_NEGATIVE_REGULATION_OF_TRANSPORT, GO_VACUOLAR_MEMBRANE               

Now we have a table for genes from the leading-edge gene sets associated
with both trajectory and trait. We merge this table with the gene-trait
p-values from MAGMA to prioritize the genes, keeping only genes with
p-value &lt; 0.05.

``` r
risk.genes <- merge(risk.genes, gene.p[,c(1,8,9)], by = "GENE") 
risk.genes <- dplyr::filter(risk.genes, P < 0.05 ) %>% dplyr::arrange(P)
head(risk.genes,3)
```

         GENE
    1 SLC30A8
    2   KCNQ1
    3    WFS1
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Gene-sets
    1                                                                                                                                                                                                                   GO_TRANSPORT_VESICLE_MEMBRANE, GO_HORMONE_TRANSPORT, GO_REGULATION_OF_PEPTIDE_SECRETION, GO_TRANSPORT_VESICLE, GO_REGULATION_OF_HORMONE_LEVELS, GO_PEPTIDE_SECRETION, GO_SIGNAL_RELEASE, GO_SECRETORY_GRANULE_MEMBRANE, HP_ADULT_ONSET, GO_METAL_ION_HOMEOSTASIS, HP_ABNORMAL_GLUCOSE_HOMEOSTASIS, GO_CELLULAR_ION_HOMEOSTASIS, GO_DIVALENT_INORGANIC_CATION_HOMEOSTASIS
    2                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  GO_REGULATION_OF_CARDIAC_MUSCLE_CONTRACTION, GO_REGULATION_OF_STRIATED_MUSCLE_CONTRACTION
    3 GO_ENDOPLASMIC_RETICULUM_UNFOLDED_PROTEIN_RESPONSE, GO_CELLULAR_RESPONSE_TO_TOPOLOGICALLY_INCORRECT_PROTEIN, GO_TRANSPORT_VESICLE_MEMBRANE, GO_PROTEIN_MATURATION, GO_RESPONSE_TO_ENDOPLASMIC_RETICULUM_STRESS, GO_INTRINSIC_COMPONENT_OF_ORGANELLE_MEMBRANE, GO_TRANSPORT_VESICLE, GO_RESPONSE_TO_TOPOLOGICALLY_INCORRECT_PROTEIN, GO_IRE1_MEDIATED_UNFOLDED_PROTEIN_RESPONSE, HP_ADULT_ONSET, GO_INTRINSIC_COMPONENT_OF_ENDOPLASMIC_RETICULUM_MEMBRANE, GO_METAL_ION_HOMEOSTASIS, HP_ABNORMAL_GLUCOSE_HOMEOSTASIS, GO_CELLULAR_ION_HOMEOSTASIS, GO_DIVALENT_INORGANIC_CATION_HOMEOSTASIS
        ZSTAT          P
    1 11.5250 4.9257e-31
    2 10.6780 6.4236e-27
    3  8.4311 1.7116e-17

We color the cells in the trajectory plot according to the expression of
the genes as follows. We load the cell metadata from [**step
1d**](https://github.com/ElkonLab/scGWAS/blob/master/vignettes/1d.md).

``` r
meta <- readRDS("meta.pancreas.RDS")
```

And the normalized, human ortholog converted matrix from [**step
1b**](https://github.com/ElkonLab/scGWAS/blob/master/vignettes/1b.md).

``` r
exp <- readRDS("cm.pancreas.RDS")
```

``` r
# genes to plot
genes.plot <- risk.genes$GENE[1:3]
print(genes.plot)
```

    [1] "SLC30A8" "KCNQ1"   "WFS1"   

``` r
#Subset the expression matrix.
exp <- exp[which(rownames(exp) %in% genes.plot),]

# scale data using Seurat
exp.scaled <- Seurat::ScaleData(exp)
exp.scaled <- t(exp.scaled)
# Floor using Seurat
exp.scaled <- apply(exp.scaled, 2, Seurat::MinMax, 
                    min = -1.5, max = 1.5)
exp.scaled <- cbind.data.frame(data.frame(cells = rownames(exp.scaled)), 
                               as.data.frame(exp.scaled))

# Merge gene expression and dimensional reduction coordinates.
exp.scaled <- merge(exp.scaled, meta, by = "cells")

for(i in genes.plot){
p <- ggplot(exp.scaled, ggplot2::aes(y = dim2 , x = dim1,
                                  color =  get(i)))  +
  geom_point(size=1) + xlab("Component 1") + ylab("Component 2") +
  ggpubr::theme_pubr()  +
  scale_color_gradient2(low ="#FF00FF", mid = "#000000", midpoint = 0,
                        high ="#FFFF00", name = i)
print(p)
}
```

![](https://github.com/ElkonLab/scGWAS/blob/master/data/pic/slc.png)
![](https://github.com/ElkonLab/scGWAS/blob/master/data/pic/kcnq1.png)
![](https://github.com/ElkonLab/scGWAS/blob/master/data/pic/wfs1.png)
